### PR TITLE
Add mailer script, main script, small tweak to pdf generating script, clean and complete requirements.txt (x2)

### PR DESCRIPTION
**Pull request by e-lemma**
Fairly large PR, title says it all.
Dockerfile coming soon.

### CODE REVIEW

File: pdf_report/mailer.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+
+import boto3
+import botocore.exceptions
+from botocore.exceptions import ClientError
+from dotenv import load_dotenv
+import psycopg2
+import psycopg2.extras
+from psycopg2.extensions import connection as DBConnection, cursor as DBCursor
+
+load_dotenv()
```
Comment (Zhi-704) [inline pdf_report/mailer.py:19]: Disagree with this being here, why not main?

File: pdf_report/mailer.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    recipient_name: str,
+    subject: str,
+) -> None:
+    """Sends an email with a PDF attachment using Amazon SES."""
+
+    body_html = f"""\
+    <html>
+    <head></head>
+    <body>
+    <h1>Good morning, {recipient_name}!</h1>
```
Comment (Zhi-704) [inline pdf_report/mailer.py:91]: Cheers for adding customnisability

File: pdf_report/mailer.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+            recipient_email,
+            e.response["Error"]["Message"],
+        )
+
+
+def send_all_emails() -> None:
+    """Main function to send the PDF attached to an email to all recipients."""
+
+    logging.basicConfig(
+        level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s"
```
Comment (Zhi-704) [inline pdf_report/mailer.py:132]: I don't know if this should be here?

File: pdf_report/mailer.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+    conn = None
+    try:
+        conn = get_connection()
+        subscribers = get_recipients(conn)
+    except Exception as e:
+        logging.error("Failed to retrieve recipients: %s", e)
+        raise Exception(f"Error retrieving recipients: {e}") from e
+    finally:
+        if conn:
```
Comment (Zhi-704) [inline pdf_report/mailer.py:150]: Good catch

File: pdf_report/pdf_report.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+def main(event, context):  # pylint: disable=unused-argument
+    """Main function to execute the:
+        1. Generation of the daily report PDFs
+        2. Sending of PDFs to all subscribers
+    """
+
+    logging.basicConfig(
+        level=logging.WARNING, format="%(asctime)s - %(levelname)s - %(message)s"
+    )
+
```
Comment (Zhi-704) [inline pdf_report/pdf_report.py:17]: This was already called inside the function

File: pdf_report/mailer.py (hunk 1 above)
Comment (e-lemma) [reply pdf_report/mailer.py:19]: It doesn't need to be in the main block as all it does is load the environment variables. Furthermore, I have a global constant that uses an environment variable. All in all, I don't think it matters too much

File: pdf_report/mailer.py (hunk 2 above)
Comment (e-lemma) [reply pdf_report/mailer.py:91]: You are welcome!

File: pdf_report/pdf_report.py (hunk 5 above)
Comment (e-lemma) [reply pdf_report/pdf_report.py:17]: Yes, but I _think_  that because it is a different file, the logger has to be configured again?

Comment (e-lemma) [reply pdf_report/pdf_report.py:17]: Actually... you might be right...

Comment (Zhi-704) [reply pdf_report/pdf_report.py:17]: I think so as well

**Pull request merged**